numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,891 @@
+"""
+Wrapper class around the ndarray object for the array API standard.
+
+The array API standard defines some behaviors differently than ndarray, in
+particular, type promotion rules are different (the standard has no
+value-based casting). The standard also specifies a more limited subset of
+array methods and functionalities than are implemented on ndarray. Since the
+goal of the array_api namespace is to be a minimal implementation of the array
+API standard, we need to define a separate wrapper class for the array_api
+namespace.
+
+The standard compliant class is only a wrapper class. It is *not* a subclass
+of ndarray.
+"""
+
+from __future__ import annotations
+
+import operator
+from enum import IntEnum
+from ._creation_functions import asarray
+from ._dtypes import _boolean_dtypes, _integer_dtypes, _floating_dtypes
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from ._types import Optional, PyCapsule, Tuple, Union, array
+
+import numpy as np
+
+class ndarray:
+    """
+    ndarray object for the array API namespace.
+
+    See the docstring of :py:obj:`np.ndarray <numpy.ndarray>` for more
+    information.
+
+    This is a wrapper around numpy.ndarray that restricts the usage to only
+    those things that are required by the array API namespace. Note,
+    attributes on this object that start with a single underscore are not part
+    of the API specification and should only be used internally. This object
+    should not be constructed directly. Rather, use one of the creation
+    functions, such as asarray().
+
+    """
+    # Use a custom constructor instead of __init__, as manually initializing
+    # this class is not supported API.
+    @classmethod
+    def _new(cls, x, /):
+        """
+        This is a private method for initializing the array API ndarray
+        object.
+
+        Functions outside of the array_api submodule should not use this
+        method. Use one of the creation functions instead, such as
+        ``asarray``.
+
+        """
+        obj = super().__new__(cls)
+        # Note: The spec does not have array scalars, only shape () arrays.
+        if isinstance(x, np.generic):
+            # Convert the array scalar to a shape () array
+            xa = np.empty((), x.dtype)
+            xa[()] = x
+            x = xa
+        obj._array = x
```

Comment:
No code outside of the array API should directly construct this class, so if this happens, it means there's a bug in the array API code. The various array API functions are all implicitly assumed to only accept inputs that they are type hinted against (i.e., they do things like `x._array` if `x` is supposed to be an array API `ndarray` object).

Reply:
> The various array API functions are all implicitly assumed to only accept inputs that they are type hinted against

Fair enough, then they should work correctly, since `array` is currently typed as `_array_api.ndarray`, I guess.  Shouldn't they run-time enforce that though?